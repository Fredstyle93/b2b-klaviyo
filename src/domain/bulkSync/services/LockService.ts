export interface LockService {
    acquireLock(key: string): void;
    releaseLock(key: string): void;
}
