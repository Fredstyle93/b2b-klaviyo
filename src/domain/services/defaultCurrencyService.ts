export class DefaultCurrencyService implements ICurrencyService {
    convert(value: number, currencyIso: string): number {
        // ***** hardcoded conversion to USD ***
        // switch (currencyIso) {
        //     case 'GBP':
        //         return value / 1.21;
        //     case 'EUR':
        //         return value / 1.07;
        //     default:
        //         return value;
        // }
        //
        // ***** API service for currencies conversion ***
        // const result = await fetch(`https://my-conversion-api-service/${value}?currency=${currencyIso}`);
        // return result.json();
        //
        // ***** no conversion *****
        return value;
    }
}
