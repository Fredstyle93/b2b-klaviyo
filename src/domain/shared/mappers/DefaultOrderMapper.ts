import { getCustomerProfileFromOrder } from '../../../utils/get-customer-profile-from-order';
import { getTypedMoneyAsNumber } from '../../../utils/get-typed-money-as-number';
import { mapAllowedProperties } from '../../../utils/property-mapper';
import { LineItem, Order } from '@commercetools/platform-sdk';
import { OrderMapper } from './OrderMapper';
import config from 'config';
import { CurrencyService } from '../services/CurrencyService';

export class DefaultOrderMapper implements OrderMapper {
    constructor(private readonly currencyService: CurrencyService) {}
    public mapCtOrderToKlaviyoEvent(order: Order, metric: string, time?: string): EventRequest {
        return {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(order),
                    metric: {
                        name: metric,
                    },
                    value: this.currencyService.convert(
                        getTypedMoneyAsNumber(order?.totalPrice),
                        order.totalPrice.currencyCode,
                    ),
                    properties: mapAllowedProperties('order', { ...order }) as any,
                    unique_id: order.id,
                    time: time ?? order.createdAt,
                },
            },
        };
    }

    public mapOrderLineToProductOrderedEvent(lineItem: LineItem, order: Order, time?: string): EventRequest {
        return {
            data: {
                type: 'event',
                attributes: {
                    profile: getCustomerProfileFromOrder(order),
                    metric: {
                        name: config.get('order.metrics.orderedProduct'),
                    },
                    value: this.currencyService.convert(
                        getTypedMoneyAsNumber(lineItem.totalPrice),
                        order.totalPrice.currencyCode,
                    ),
                    properties: { ...lineItem },
                    unique_id: lineItem.id,
                    time: time ?? order.createdAt,
                },
            },
        };
    }
}
