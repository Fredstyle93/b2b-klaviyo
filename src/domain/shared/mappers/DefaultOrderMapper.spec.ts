import { DefaultOrderMapper } from "./DefaultOrderMapper";
import { mock } from "jest-mock-extended";
import { DummyCurrencyService } from "../services/dummyCurrencyService";
import { sampleOrderCreatedMessage } from "../../../test/testData/orderData";

const mockCurrencyService = mock<DummyCurrencyService>();
mockCurrencyService.convert.mockImplementation((value, currency) => value)
const orderMapper = new DefaultOrderMapper(mockCurrencyService)
describe("map CT order to Klaviyo event", () => {
  it("should map a commercetools order with a given metric to a klaviyo event", () => {

    const klaviyoEvent = orderMapper.mapCtOrderToKlaviyoEvent(sampleOrderCreatedMessage.order, "someMetric");
    expect(klaviyoEvent).toMatchSnapshot();
  })
})
describe("map CT order line to klaviyo event", () => {
  it("should map a commercetools order line to a klaviyo Order Product event", () => {

    const klaviyoEvent = orderMapper.mapOrderLineToProductOrderedEvent(
        {
            discountedPricePerQuantity: [],
            lineItemMode: "",
            perMethodTaxRate: [],
            price: { id: "someId", value: { type: 'centPrecision', centAmount: 1300, currencyCode: 'USD', fractionDigits: 2}  },
            priceMode: "",
            productId: '',
            productType: { id: "id", typeId: "product-type"},
            quantity: 0,
            state: [],
            taxedPricePortions: [],
            variant: { id: 1},
            id: '123-123-123',
            totalPrice: { type: 'centPrecision', centAmount: 1300, currencyCode: 'USD', fractionDigits: 2 },
            name: {
                en: 'Test product',
            },
        },
        sampleOrderCreatedMessage.order,
    );
    expect(klaviyoEvent).toMatchSnapshot();
  })
})
