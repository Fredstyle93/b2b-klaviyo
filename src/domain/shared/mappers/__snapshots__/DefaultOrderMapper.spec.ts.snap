// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map CT order line to klaviyo event should map a commercetools order line to a klaviyo Order Product event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "Ordered Product",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "discountedPricePerQuantity": [],
        "id": "123-123-123",
        "lineItemMode": "",
        "name": {
          "en": "Test product",
        },
        "perMethodTaxRate": [],
        "price": {
          "id": "someId",
          "value": {
            "centAmount": 1300,
            "currencyCode": "USD",
            "fractionDigits": 2,
            "type": "centPrecision",
          },
        },
        "priceMode": "",
        "productId": "",
        "productType": {
          "id": "id",
          "typeId": "product-type",
        },
        "quantity": 0,
        "state": [],
        "taxedPricePortions": [],
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "variant": {
          "id": 1,
        },
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "123-123-123",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT order to Klaviyo event should include/exclude/map custom fields in a commercetools order and send them with the klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "includedFieldAfterMap": true,
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT order to Klaviyo event should map a commercetools order with a given metric to a klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT refunded order to Klaviyo event should include/exclude/map custom fields in a commercetools order and send them with the klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "includedFieldAfterMap": true,
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;

exports[`map CT refunded order to Klaviyo event should map a commercetools refunded order with a given metric to a klaviyo event 1`] = `
{
  "data": {
    "attributes": {
      "metric": {
        "name": "someMetric",
      },
      "profile": {
        "$email": "test@klaviyo.com",
        "$id": "123-123-123",
      },
      "properties": {
        "createdAt": "2023-01-27T15:00:00.000Z",
        "customItems": [],
        "customerEmail": "test@klaviyo.com",
        "customerId": "123-123-123",
        "items": [],
        "lastModifiedAt": "2023-01-27T15:00:00.000Z",
        "orderId": "3456789",
        "orderState": "Open",
        "totalPrice": {
          "centAmount": 1300,
          "currencyCode": "USD",
          "fractionDigits": 2,
          "type": "centPrecision",
        },
        "version": 24,
      },
      "time": "2023-01-27T15:00:00.000Z",
      "unique_id": "3456789",
      "value": 13,
    },
    "type": "event",
  },
}
`;
